persistence/sql/sql_process_store_common.go

```diff
@@ -78,3 +84,89 @@ func insertImmediateTask(
 
 	return tx.InsertImmediateTask(ctx, immediateTaskRow)
 }
+
+// publishToLocalQueue inserts len(valid_messages) rows into xdb_sys_local_queue,
+// and inserts only one row into xdb_sys_immediate_tasks with all the dedupIds for these messages.
+// publishToLocalQueue returns (HasNewImmediateTask, error).
+func (p sqlProcessStoreImpl) publishToLocalQueue(
+	ctx context.Context, tx extensions.SQLTransaction, processExecutionId uuid.UUID, messages []xdbapi.LocalQueueMessage) (bool, error) {
+
+	var localQueueMessageInfo []persistence.LocalQueueMessageInfoJson
+
+	for _, message := range messages {
+		dedupId := uuid.MustNewUUID()
+
+		// dealing with user-customized dedupId
+		if message.GetDedupId() != "" {
+			dedupId2, err := uuid.ParseUUID(message.GetDedupId())
+			if err != nil {
+				return false, err
+			}
+
+			// need to check if this message has been published before
+			existingMessage, err := p.session.SelectLocalQueue(ctx, processExecutionId, message.GetQueueName(), dedupId2)
```

Comment:
This checking can be done via the insert query -- it will returns error on conflict